README.md
# mc-wallet
[![Latest Release](https://img.shields.io/github/v/release/Ubunfu/mc-wallet)](https://github.com/Ubunfu/mc-wallet/releases)
[![codecov](https://codecov.io/gh/Ubunfu/mc-wallet/branch/master/graph/badge.svg?token=258GHULYE1)](https://codecov.io/gh/Ubunfu/mc-wallet)
[![CircleCI](https://img.shields.io/circleci/build/github/Ubunfu/mc-wallet?logo=circleci)](https://app.circleci.com/pipelines/github/Ubunfu/mc-wallet)
![Contrubutors](https://img.shields.io/github/contributors/Ubunfu/mc-wallet?color=blue)
![Last Commit](https://img.shields.io/github/last-commit/Ubunfu/mc-wallet)

Processes requests for payments and charges to user wallets.

This service runs as an AWS lambda function.

## Configuration
### IAM Role
AWS's standard IAM role for Lambda micro services is plenty sufficient. The only access that is required is read/write for DynamoDB, and write to CloudWatch logs.

* AWSLambdaMicroserviceExecutionRole
* AWSLambdaBasicExecutionRole

### Environment Variables
| Parameter          | Description                                                                       | Default | Required? |
|--------------------|-----------------------------------------------------------------------------------|---------|-----------|
| TABLE_WALLETS      | The name of a DynamoDB table containing user wallets.                             | n/a     | Yes       |
| LOGGER_ENABLED     | Boolean value controlling writing of logs. Useful to turn off for test execution. | n/a     | Yes       |

CHANGELOG.md
# Changelog
This project attempts to conform to [SemVer](https://semver.org/) guidelines whenever possible.

## [v1.0.0]
* [Adding DB to serverless stack](https://github.com/Ubunfu/mc-wallet/pull/16)
* [Adding Circle Orb to post build notifications to Discord](https://github.com/Ubunfu/mc-wallet/pull/17)

## [v0.4.1]
* [Preventing stealing via negative payments](https://github.com/Ubunfu/mc-wallet/pull/14)

## [v0.4.0]
* [Adopting serverless framework](https://github.com/Ubunfu/mc-wallet/pull/12)

## [v0.3.0]
* [Adding UML diagrams](https://github.com/Ubunfu/mc-wallet/pull/8)
* [Removing sensitive data from logs](https://github.com/Ubunfu/mc-wallet/pull/9)
* [Fixing changelog version](https://github.com/Ubunfu/mc-wallet/pull/10)

## [v0.2.0]
* [Implementing get wallet API](https://github.com/Ubunfu/mc-wallet/pull/5)
* [Adding debug logging](https://github.com/Ubunfu/mc-wallet/pull/6)

## [v0.1.1]
* [Adding API specification](https://github.com/Ubunfu/mc-wallet/pull/3)
* Adding changelog

## [v0.1.0]
* Initial implementation

test/test.js
describe('When a test is run', function() {
    it('Should pass')
});

test/indexTest.js
const index = require('../src/index.js');
const expect = require('chai').expect;
const sinon = require('sinon');
const paymentService = require('../src/service/paymentService.js');
const dbService = require('../src/service/dbService.js');
const errorMessageEnum = require('../src/enums/errorMessageEnum.js');
const paymentServiceErrorEnum = require('../src/enums/paymentServiceErrorEnum.js');

const EVENT_PAYMENT = {
    requestContext: {
        routeKey: 'POST /pay'
    },
    body: '{"player": "player1", "amount": 100}'
}
const EVENT_CHARGE = {
    requestContext: {
        routeKey: 'POST /charge'
    },
    body: '{"player": "player1", "amount": 100}'
}
const EVENT_GET_WALLET = {
    requestContext: {
        routeKey: 'GET /wallet'
    },
    queryStringParameters: {
        id: 'player1'
    }
}
const A_WALLET = {
    WalletId: 'player',
    Balance: 100
}

describe('index: When receives a get wallet request', function() {
    describe('And wallet not found', function() {
        it('Returns HTTP 404', async function() {
            const dbServiceMock = sinon.stub(dbService, "getWallet")
                .returns({});
            const walletResp = await index.handler(EVENT_GET_WALLET);
            expect(walletResp.statusCode).to.be.equal('404');
            expect(JSON.parse(walletResp.body).error).to.be.equal(errorMessageEnum.ERROR_GET_WALLET_FAILED);
            expect(JSON.parse(walletResp.body).errorDetail).to.be.equal(paymentServiceErrorEnum.ERROR_WALLET_NOT_FOUND);
            dbServiceMock.restore();
        });
    });
    describe('And some other error happens', function() {
        it('Returns HTTP 500', async function() {
            const dbServiceMock = sinon.stub(dbService, "getWallet")
                .throws('errorName', errorMessageEnum.ERROR_GET_WALLET_FAILED);
            const walletResp = await index.handler(EVENT_GET_WALLET);
            expect(walletResp.statusCode).to.be.equal('500');
            expect(JSON.parse(walletResp.body).error).to.be.equal(errorMessageEnum.ERROR_GET_WALLET_FAILED);
            expect(JSON.parse(walletResp.body).errorDetail).to.be.equal(errorMessageEnum.ERROR_GET_WALLET_FAILED);
            dbServiceMock.restore();
        });
    });
    describe('And DB returns a wallet', function() {
        it('Returns the wallet', async function() {
            const dbServiceMock = sinon.stub(dbService, "getWallet")
                .returns(A_WALLET);
            const walletResp = await index.handler(EVENT_GET_WALLET);
            expect(walletResp.statusCode).to.be.equal('200');
            expect(JSON.parse(walletResp.body)).to.be.deep.equal(A_WALLET);
            dbServiceMock.restore();
        });
    });
});

describe('index: When receives wallet payment request', function() {
    describe('And payment service succeeds', function() {
        let handlerResp, paymentServiceMock = null;
        beforeEach(async function() {
            paymentServiceMock = sinon.stub(paymentService, "pay")
                .returns({});
            handlerResp = await index.handler(EVENT_PAYMENT);
        });
        afterEach(function() {
            paymentServiceMock.restore();
        });
        it('Should call payment service with player and amount', function() {
            expect(paymentServiceMock.calledOnceWith("player1", 100)).to.be.true;
        });
        it('Should return HTTP 200', function() {
            expect(handlerResp.statusCode).to.be.equal('200');
        });
    });
    describe('And payment service throws non-positive amount error', function() {
        it('Returns HTTP 400', async function() {
            const paymentServiceMock = sinon.stub(paymentService, "pay")
                .throws('errorName', paymentServiceErrorEnum.ERROR_NON_POSITIVE_AMOUNT);
            const handlerResp = await index.handler(EVENT_PAYMENT);
            expect(handlerResp.statusCode).to.be.equal('400');
            paymentServiceMock.restore();
        });
    });
    describe('And payment service fails', function() {
        let handlerResp, paymentServiceMock = null;
        beforeEach(async function() {
            paymentServiceMock = sinon.stub(paymentService, "pay")
                .throws('errorName', 'errorMessage!');
            handlerResp = await index.handler(EVENT_PAYMENT);
        });
        afterEach(function() {
            paymentServiceMock.restore();
        });
        it('Should return HTTP 500', function() {
            expect(handlerResp.statusCode).to.be.equal('500');
        });
        it('Should return error code', function() {
            const respBody = JSON.parse(handlerResp.body);
            expect(respBody.error).to.be.equal(errorMessageEnum.ERROR_PAY_WALLET_FAILED);
            expect(respBody.errorDetail).to.be.equal('errorMessage!');
        });
    });
});

describe('index: When receives wallet charge request', function() {
    describe('And payment service succeeds', function() {
        let handlerResp, paymentServiceMock = null;
        beforeEach(async function() {
            paymentServiceMock = sinon.stub(paymentService, "charge")
                .returns({});
            handlerResp = await index.handler(EVENT_CHARGE);
        });
        afterEach(function() {
            paymentServiceMock.restore();
        });
        it('Should call payment service with player and amount', function() {
            expect(paymentServiceMock.calledOnceWith("player1", 100));
        });
        it('Should return HTTP 200', function() {
            expect(handlerResp.statusCode).to.be.equal('200');
        });
    });
    describe('And payment service throws negative amount error', function() {
        it('Returns HTTP 400', async function() {
            const paymentServiceMock = sinon.stub(paymentService, "charge")
                .throws('errorName', paymentServiceErrorEnum.ERROR_NON_POSITIVE_AMOUNT);
            const handlerResp = await index.handler(EVENT_CHARGE);
            expect(handlerResp.statusCode).to.be.equal('400');
            paymentServiceMock.restore();
        });
    });
    describe('And payment service fails because wallet missing', function() {
        let handlerResp, paymentServiceMock = null;
        beforeEach(async function() {
            paymentServiceMock = sinon.stub(paymentService, "charge")
                .throws('errorName', paymentServiceErrorEnum.ERROR_WALLET_NOT_FOUND);
            handlerResp = await index.handler(EVENT_CHARGE);
        });
        afterEach(function() {
            paymentServiceMock.restore();
        });
        it('Should return HTTP 404', function() {
            expect(handlerResp.statusCode).to.be.equal('404');
        });
        it('Should return error code', function() {
            const respBody = JSON.parse(handlerResp.body);
            expect(respBody.error).to.be.equal(errorMessageEnum.ERROR_CHARGE_WALLET_FAILED);
            expect(respBody.errorDetail).to.be.equal(paymentServiceErrorEnum.ERROR_WALLET_NOT_FOUND);
        });
    });
    describe('And payment service fails otherwise', function() {
        let handlerResp, paymentServiceMock = null;
        beforeEach(async function() {
            paymentServiceMock = sinon.stub(paymentService, "charge")
                .throws('errorName', 'unexpected error');
            handlerResp = await index.handler(EVENT_CHARGE);
        });
        afterEach(function() {
            paymentServiceMock.restore();
        });
        it('Should return HTTP 500', function() {
            expect(handlerResp.statusCode).to.be.equal('500');
        });
        it('Should return error code', function() {
            const respBody = JSON.parse(handlerResp.body);
            expect(respBody.error).to.be.equal(errorMessageEnum.ERROR_CHARGE_WALLET_FAILED);
            expect(respBody.errorDetail).to.be.equal('unexpected error');
        });
    });
});

describe('index: Receives unknown request', function() {
    
    it('Should return HTTP 404', async function() {
        const handlerResp = await index.handler({
            requestContext: {
                http: {
                    path: '/notfound'
                }
            }
        });
        expect(handlerResp.statusCode).to.be.equal('404');
    });
})


test/util/loggerTest.js
const logger = require('../../src/util/logger.js');
const sinon = require('sinon');
const expect = require('chai').expect;

describe('logger: When logger is enabled', function() {
    it('Should write log', function() {
        process.env.LOGGER_ENABLED = true;
        const consoleStub = sinon.stub(console, "log");
        logger.log("message");
        expect(consoleStub.calledOnceWithExactly("message")).to.be.true;
        consoleStub.restore();
    });
});

describe('logger: When logger is disabled', function() {
    it('Should not write log', function() {
        process.env.LOGGER_ENABLED = false;
        const consoleStub = sinon.stub(console, "log");
        logger.log("message");
        expect(consoleStub.callCount).to.be.equal(0);
    })
});

src/service/dbService.js
const { log } = require("../util/logger");

async function getWallet(docClient, player) {
    const params = {
        TableName: process.env.TABLE_WALLETS,
        Key: {
            "WalletId": player
        }
    };
    try {
        log('Query params: ' + JSON.stringify(params));
        const theWallet = await docClient.get(params).promise();
        log('DB Returned: ' + JSON.stringify(theWallet));
        if (theWallet.Item == undefined) {
            return {};
        }
        return theWallet.Item;
    } catch (err) {
        log(err);
        throw Error('db_error');
    }
}

async function updateWallet(docClient, player, amount) {
    const params = {
        TableName: process.env.TABLE_WALLETS,
        Key:{
            "WalletId": player
        },
        UpdateExpression: "set Balance = Balance + :val",
        ExpressionAttributeValues: {
            ":val": amount
        },
        ReturnValues: "UPDATED_NEW"
    };
    try {
        log('Query params: ' + JSON.stringify(params));
        await docClient.update(params).promise();
    } catch (err) {
        log(err);
        throw Error('db_error');
    }

}

async function createWallet(docClient, player) {
    const params = {
        TableName: process.env.TABLE_WALLETS,
        Item: {
            "WalletId": player,
            "Balance": 0
        }
    };
    try {
        log('Query params: ' + JSON.stringify(params));
        await docClient.put(params).promise();
    } catch (err) {
        log(err);
        throw Error('db_error');
    }
}

exports.getWallet = getWallet;
exports.updateWallet = updateWallet;
exports.createWallet = createWallet;

src/service/paymentService.js
const dbService = require('../service/dbService.js');
const AWS = require('aws-sdk');
const { log } = require('../util/logger.js');
const docClient = new AWS.DynamoDB.DocumentClient();
const paymentServiceErrorEnum = require('../enums/paymentServiceErrorEnum.js');

async function pay(player, amount) {

    if (amount <= 0) {
        throw Error(paymentServiceErrorEnum.ERROR_NON_POSITIVE_AMOUNT);
    }
    
    let walletResp;
    try {
        walletResp = await dbService.getWallet(docClient, player);
    } catch (err) {
        log('error retrieving wallet: ' + JSON.stringify(err));
        throw Error(paymentServiceErrorEnum.ERROR_GET_WALLET_FAILED);
    }
    
    if(walletResp.WalletId == undefined) {
        try {
            await dbService.createWallet(docClient, player);
        } catch (err) {
            log('error creating wallet: ' + JSON.stringify(err));
            throw Error(paymentServiceErrorEnum.ERROR_CREATE_WALLET_FAILED);
        }
    }
    
    try {
        await dbService.updateWallet(docClient, player, amount);
    } catch (err) {
        log('error updating wallet: ' + JSON.stringify(err));
        throw Error(paymentServiceErrorEnum.ERROR_UPDATE_WALLET_FAILED);
    }
}

async function charge(player, amount) {
    
    if (amount <= 0) {
        throw Error(paymentServiceErrorEnum.ERROR_NON_POSITIVE_AMOUNT);
    }
    
    let walletResp;
    try {
        walletResp = await dbService.getWallet(docClient, player);
    } catch (err) {
        log('error retrieving wallet: ' + JSON.stringify(err));
        throw Error(paymentServiceErrorEnum.ERROR_GET_WALLET_FAILED);
    }
    
    if(walletResp.WalletId == undefined) {
        throw Error(paymentServiceErrorEnum.ERROR_WALLET_NOT_FOUND);
    }

    if (walletResp.Balance < amount) {
        throw Error(paymentServiceErrorEnum.ERROR_INSUFFICIENT_FUNDS);
    }

    try {
        await dbService.updateWallet(docClient, player, -amount);
    } catch (err) {
        log('error updating wallet: ' + JSON.stringify(err));
        throw Error(paymentServiceErrorEnum.ERROR_UPDATE_WALLET_FAILED);
    }
}

exports.pay = pay;
exports.charge = charge;

src/index.js
require('dotenv').config();
const { log } = require('./util/logger.js');
const paymentService = require('./service/paymentService.js');
const dbService = require('./service/dbService.js');
const AWS = require('aws-sdk');
const docClient = new AWS.DynamoDB.DocumentClient();
const errorMessageEnum = require('./enums/errorMessageEnum.js');
const paymentServiceErrorEnum = require('./enums/paymentServiceErrorEnum.js');

exports.handler = async (event, context) => {
    logEvent(event);

    let body;
    let statusCode = '200';
    const headers = {
        'Content-Type': 'application/json',
    };

    if (event.requestContext.routeKey == 'POST /pay') {
        const player = JSON.parse(event.body).player;
        const amount = JSON.parse(event.body).amount;
        try {
            await paymentService.pay(player, amount);
        } catch (err) {
            if (err.message == paymentServiceErrorEnum.ERROR_NON_POSITIVE_AMOUNT) {
                statusCode = '400';
            } else {
                statusCode = '500';
            }
            body = {
                error: errorMessageEnum.ERROR_PAY_WALLET_FAILED,
                errorDetail: err.message
            }
        }
    } else if (event.requestContext.routeKey == 'POST /charge') {
        const player = JSON.parse(event.body).player;
        const amount = JSON.parse(event.body).amount;
        try {
            await paymentService.charge(player, amount);
        } catch (err) {
            if (err.message == paymentServiceErrorEnum.ERROR_WALLET_NOT_FOUND) {
                statusCode = '404';
            } else if (err.message == paymentServiceErrorEnum.ERROR_NON_POSITIVE_AMOUNT) {
                statusCode = '400';
            } else {
                statusCode = '500';
            }
            body = {
                error: errorMessageEnum.ERROR_CHARGE_WALLET_FAILED,
                errorDetail: err.message
            }
        }
    } else if (event.requestContext.routeKey == 'GET /wallet') {
        const walletId = event.queryStringParameters.id;
        try {
            const walletResp = await dbService.getWallet(docClient, walletId);
            if (walletResp.WalletId == undefined) {
                statusCode = '404';
                body = {
                    error: errorMessageEnum.ERROR_GET_WALLET_FAILED,
                    errorDetail: paymentServiceErrorEnum.ERROR_WALLET_NOT_FOUND
                }
            } else {
                body = walletResp;
            }
        } catch (err) {
            statusCode = '500';
            body = {
                error: errorMessageEnum.ERROR_GET_WALLET_FAILED,
                errorDetail: err.message
            }
        }
    } else {
        statusCode = '404'
    }

    body = JSON.stringify(body);
    return {
        statusCode,
        body,
        headers,
    };
};

async function logEvent(event) {
    if (event.headers) {
        event.headers.host = "XXX";
    }
    if (event.requestContext) {
        event.requestContext.domainName = "XXX";
    }
    await log(JSON.stringify(event, null, 2));
}